### PR TITLE
!htp,jso,xml #18064 remove implicit parameters of Unmarshaller instances

### LINKED ISSUE #18064: Implicit Materializer is a usability problem
@sirthias removed `ExecutionContext` implicits from marshalling / unmarshalling which already improved usability somewhat. The need to have a `Materializer` in scope for many often used Unmarshallers like json, ByteString, or StrictForm proves as a remaining stumbling block not only for new users (e.g. see here: https://groups.google.com/d/topic/akka-user/Vhm_zkTajJk/discussion).

I suggest to do a similar transformation for the `Materializer` value to remove it from the implicit values and move it into the `Unmarshaller` itself and provide it from the `RequestContext` at runtime.

### PR DESCRIPTION
**Pull request by jrudolph**
Fixes #18064.

/cc @sirthias

### CODE REVIEW

**Comment by jrudolph** [comment]
Error lines in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/3279/:

```
   926 /mnt/akka/jenkinsakka/localhome-a4/workspace/pr-validator-per-commit-jenkins/akka-docs-dev/rst/scala/code/docs/http/scaladsl/UnmarshalSpec.scala:18: could not find implicit value for parameter mat: akka.stream.Materializer
   927     val intFuture = Unmarshal("42").to[Int]
   928                                       ^
  1003 /mnt/akka/jenkinsakka/localhome-a4/workspace/pr-validator-per-commit-jenkins/akka-docs-dev/rst/scala/code/docs/stream/CompositionDocSpec.scala:209: value p in class MyClass cannot be accessed in MyClass
  1004     case class MyClass(private val p: Promise[Unit], conn: OutgoingConnection) {
  1005                ^
  1021 two errors found
  1022 (akka-docs-dev/test:compile) Compilation failed
  1023 Total time: 440 s, completed Jul 23, 2015 7:45:41 PM
```

**Comment by jrudolph** [comment]
I have seen this error for a few times now but it seems to come and go:

```
[error] /home/johannes/git/opensource/akka-streams/akka-docs-dev/rst/scala/code/docs/stream/CompositionDocSpec.scala:209: value p in class MyClass cannot be accessed in MyClass
[error]     case class MyClass(private val p: Promise[Unit], conn: OutgoingConnection) {
[error]                ^
```

(I fixed the other one and now it compiles for me.)

**Comment by jrudolph** [comment]
Error lines in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/3281/:

```
 10692 Failed: Total 867, Failed 1, Errors 0, Passed 866, Pending 22
 10693 Failed tests:
 10694 	akka.stream.impl.fusing.InterpreterStressSpec
 14053 (akka-stream-tests-experimental/test:test) sbt.TestsFailedException: Tests unsuccessful
 14054 Total time: 868 s, completed Jul 24, 2015 10:49:51 AM
```

File: akka-http-marshallers-scala/akka-http-spray-json/src/main/scala/akka/http/scaladsl/marshallers/sprayjson/SprayJsonSupport.scala [hunk 1]
```diff
@@ -16,11 +16,11 @@ import spray.json._
  * A trait providing automatic to and from JSON marshalling/unmarshalling using an in-scope *spray-json* protocol.
  */
 trait SprayJsonSupport {
-  implicit def sprayJsonUnmarshallerConverter[T](reader: RootJsonReader[T])(implicit mat: Materializer): FromEntityUnmarshaller[T] =
-    sprayJsonUnmarshaller(reader, mat)
-  implicit def sprayJsonUnmarshaller[T](implicit reader: RootJsonReader[T], mat: Materializer): FromEntityUnmarshaller[T] =
+  implicit def sprayJsonUnmarshallerConverter[T](reader: RootJsonReader[T]): FromEntityUnmarshaller[T] =
```
Comment (ktoso) [inline akka-http-marshallers-scala/akka-http-spray-json/src/main/scala/akka/http/scaladsl/marshallers/sprayjson/SprayJsonSupport.scala]: very nice!

File: akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/FormFieldDirectivesSpec.scala [hunk 2]
```diff
@@ -12,6 +12,9 @@ import akka.http.scaladsl.model._
 import MediaTypes._
 
 class FormFieldDirectivesSpec extends RoutingSpec {
+  // FIXME: unfortunately, it has make a come back, this time it's reproducible ...
+  import akka.http.scaladsl.server.directives.FormFieldDirectives.FieldMagnet
```
Comment (ktoso) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/FormFieldDirectivesSpec.scala]: What's the issue here? Does it need a ticket for it?

Comment (jrudolph) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/FormFieldDirectivesSpec.scala]: It's the old problem where a compilation error happens depending on the order in which files are compiled. I thought I fixed it in #17552 but now it came back and I haven't yet found the reason.

Here is some background: https://groups.google.com/d/topic/scala-language/KZhcHb6e8DI/discussion 

Yes, I'll create a ticket.

**Comment by ktoso** [comment]
LGTM, error seems unrelated to this change

File: akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/FormFieldDirectivesSpec.scala (hunk 2 above)
Comment (ktoso) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/FormFieldDirectivesSpec.scala]: Thanks, I'll read up on it!

Comment (jrudolph) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/FormFieldDirectivesSpec.scala]: https://github.com/akka/akka/issues/18070

**Pull request merged**